### ISSUE TITLE
change in `state` for terminating event when using `BacksolveAdjoint()`

### ISSUE THREAD

**Issue by snehjp2** [patrick-kidger/diffrax#294]
Hi! I am confused on what internal changes to the `state` are made when using `BacksolveAdjoint()`. I am solving a Set of 279 (but in general, N) coupled ODEs. My particular implementation has a discrete terminating event. In the scenario in which I'm not using adjoint states, my terminating event function seeks to stop the solver once a minimum threshold on the sum of $N-1$ solutions is achieved.

```
def condition_fn(state, **kwargs):
    del kwargs
    OMEGA_M_MIN = 1e-4
    Omega_m = jnp.sum(state.y[1:])
    condition = Omega_m < OMEGA_M_MIN
    return lax.cond(condition, lambda _: True, lambda _: False, None)
```

where the output of `state.y` is `Traced<ShapedArray(float64[279])>with<DynamicJaxprTrace(level=4/0)>`, and $279 = N$ as expected. After incorporating `BacksolveAdjoint()` in my `diffeqsolve`, given here:

```
sol = diffeqsolve(equation, solver, t0=t_eval[0], t1=t_eval[-1], dt0=0.01, y0=self.y0,
                        args=(self.Gamma_0,), max_steps=int(1e5), saveat=saveat,
                        stepsize_controller=stepsize_controller, discrete_terminating_event=event
                        , adjoint=BacksolveAdjoint())
````

the structure of the `state` is quite different than before. Here's the structure of `state.y`

```
Traced<ShapedArray(float64[279])>with<DynamicJaxprTrace(level=4/1)>
Traced<ShapedArray(float64[279])>with<DynamicJaxprTrace(level=4/1)>
(Traced<ShapedArray(float64[279])>with<DynamicJaxprTrace(level=3/1)>, Traced<ShapedArray(float64[279])>with<DynamicJaxprTrace(level=3/1)>, (Traced<ShapedArray(float64[278])>with<DynamicJaxprTrace(level=3/1)>,), WrapTerm(term=ODETerm(vector_field=None), direction=None))
```

I'm not sure how to properly interpret this, and how to restructure my terminating event function to handle this?

**Comment by patrick-kidger** [comment]
Ah, this looks like a bug. It looks like the terminating event is erroneously being passed to the backward pass. This shouldn't be necessary at all -- termination is a thing we only do on the forward pass; the backward pass just needs to integrate backwards from the time at which we terminated.

So to make this work, it looks like we would need to: (a) not pass the event to the backward pass at all, and (b) record the time at which we terminated, and pass that to the backward pass.

For now though, I'm just going to have this explicitly raise a `NotImplementedError`. There's almost no use-case for `BacksolveAdjoint`, after all.

**Comment by snehjp2** [comment]
Hi Patrick,

Thanks for the quick response! I am a bit confused -- I thought the motivation behind `BackSolveAdjoint` was memory efficiency? I only chose to implement it due to some OOM errors while scaling up this code. Do you have any other tips for memory management besides `BackSolveAdjoint`?

**Comment by patrick-kidger** [comment]
If you're looking to reduce memory usage then `RecursiveCheckpointAdjoint(checkpoints=...)`, with an explicit number of checkpoints (as large as possible) is usually the best option.

`BacksolveAdjoint` got a lot of press a few years ago, but it's mostly a terrible method, as it gives inaccurate approximations to the gradients. In contrast, checkpointing gives you memory efficiency without sacrificing that.
